### PR TITLE
ENH: [ArrowStringArray] Enable the string methods for the arrow-backed StringArray

### PR DESCRIPTION
**Pull request by simonjayhawkins**
from https://github.com/pandas-dev/pandas/issues/35169#issuecomment-712883859

> Enable the string methods for the arrow-backed StringArray. This might also need some additional changes in the string accessor code (eg to dispatch extract to the underlying array as well)

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -1110,6 +1110,7 @@ _TYPE_MAP = {
     "complex128": "complex",
     "c": "complex",
     "string": "string",
+    "arrow_string": "string",
```
Comment (simonjayhawkins) [inline pandas/_libs/lib.pyx:1113]: PR as draft. This will be resolved in a follow-up to #40679. https://github.com/pandas-dev/pandas/pull/40679#issuecomment-809642594

the inference is needed so that the str accessor works.

File: pandas/conftest.py [hunk 2]
```diff
@@ -1131,6 +1131,26 @@ def string_dtype(request):
...
+    Parametrized fixture for string dtypes.
+
+    * 'string'
+    * 'arrow_string'
+    """
+    from pandas.core.arrays.string_arrow import ArrowStringDtype  # noqa: F401
+
+    return request.param
+
+
```
Comment (simonjayhawkins) [inline pandas/conftest.py:1153]: This fixture is dup from  #40679 so not part of this PR

File: pandas/core/arrays/string_arrow.py [hunk 3]
```diff
@@ -680,3 +685,64 @@ def value_counts(self, dropna: bool = True) -> Series:
             raise NotImplementedError("yo")
 
         return Series(counts, index=index).astype("Int64")
+
+    # ------------------------------------------------------------------------
+    # String methods interface
+
+    _str_na_value = ArrowStringDtype.na_value
+
+    def _str_map(self, f, na_value=None, dtype: Dtype | None = None):
```
Comment (simonjayhawkins) [inline pandas/core/arrays/string_arrow.py:694]: moreless cut and paste from StringArray.

File: pandas/core/strings/accessor.py [hunk 4]
```diff
@@ -316,7 +317,7 @@ def cons_row(x):
             # This is a mess.
             dtype: Optional[str]
             if self._is_string and returns_string:
-                dtype = "string"
+                dtype = self._orig.dtype
```
Comment (simonjayhawkins) [inline pandas/core/strings/accessor.py:320]: special case here for partition/rpartition/split where the array method returns an object array

File: pandas/tests/arrays/string_/test_string.py [hunk 5]
```diff
@@ -91,17 +91,11 @@ def test_setitem_with_scalar_string(dtype):
 @pytest.mark.parametrize(
     "input, method",
     [
-        (["a", "b", "c"], operator.methodcaller("capitalize")),
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/string_/test_string.py:94]: looks like dup. maybe there is a reason

File: pandas/tests/strings/test_string_array.py [hunk 6]
```diff
@@ -73,33 +81,38 @@ def test_string_array_numeric_integer_array(method, expected):
         ("isdigit", [False, None, True]),
         ("isalpha", [True, None, False]),
         ("isalnum", [True, None, True]),
-        ("isdigit", [False, None, True]),
```
Comment (simonjayhawkins) [inline pandas/tests/strings/test_string_array.py:76]: another dup

File: pandas/tests/arrays/string_/test_string.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/string_/test_string.py:94]: I suppose this can indeed be removed

File: pandas/tests/strings/test_string_array.py [hunk 7]
```diff
@@ -11,14 +11,22 @@
...
-def test_string_array(any_string_method):
+def test_string_array(nullable_string_dtype, any_string_method, request):
     method_name, args, kwargs = any_string_method
     if method_name == "decode":
         pytest.skip("decode requires bytes.")
 
+    if nullable_string_dtype == "arrow_string" and method_name in {
+        "extract",
+        "extractall",
+    }:
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/tests/strings/test_string_array.py:22]: I think the xfail is fine for now

File: pandas/tests/strings/test_string_array.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/tests/strings/test_string_array.py:76]: Maybe can add "isnumeric" instead, which doesn't seem to be tested

**Review by jorisvandenbossche — COMMENTED** [review]
Generally looks good to me, a few comments

File: pandas/core/arrays/string_arrow.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/string_arrow.py:694]: indeed. I _think_ the de-duplication is better as an immediate follow-up to keep changes to existing code paths (i.e. StringArray) in a separate PR and keep this one scoped to just additions.

File: pandas/core/arrays/string_arrow.py [hunk 8]
```diff
@@ -43,6 +47,7 @@
     check_array_indexer,
     validate_indices,
 )
+from pandas.core.strings.object_array import ObjectStringArrayMixin
```
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/core/arrays/string_arrow.py:50]: yep. This PR just gets the string methods working for the existing tests we have StringArray. So just converting to object and not using native pyarrow functions yet.

File: pandas/core/arrays/string_arrow.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/string_arrow.py:694]: either is fine for me

File: pandas/tests/strings/test_string_array.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/tests/strings/test_string_array.py:22]: agreed. keeps the scope of this PR limited.

File: pandas/tests/strings/test_string_array.py (hunk 6 above)
Comment (simonjayhawkins) [reply pandas/tests/strings/test_string_array.py:76]: sure. "isnumeric" and other "is" functions are tested to some degree by _any_string_method fixture used in test_string_array

File: pandas/core/arrays/string_arrow.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/string_arrow.py:694]: > eg move it a common helper function or mixin

or have a common base class. https://github.com/pandas-dev/pandas/issues/35169#issuecomment-712372349

so better as a follow-on to allow for more discussion

**Comment by jorisvandenbossche** [comment]
@simonjayhawkins is this still draft? (it seems ready to me)

**Comment by simonjayhawkins** [comment]
this one goes in after #40725, https://github.com/pandas-dev/pandas/pull/40708#discussion_r605101764, over 80% complete on the benchmarks.

**Comment by jorisvandenbossche** [comment]
To be able to move forward with the different PRs, I think it is perfectly fine to merge this with the ` "arrow_string": "string",` in infer_dtype (which *is* the name of the dtype on current master), and then this can be updated when the name gets changed or in #40725

**Comment by jorisvandenbossche** [comment]
@simonjayhawkins this is ready for review/merge now?

**Comment by simonjayhawkins** [comment]
hopefully, will mark as ready for review once ci is green.

**Comment by simonjayhawkins** [comment]
@jorisvandenbossche green

File: pandas/core/arrays/string_arrow.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/arrays/string_arrow.py:48]: Can you add a small comment about that (eg below where the class is created) about why `ObjectStringArrayMixin` is mixed in?

**Review by jorisvandenbossche — APPROVED** [review]
Can you do one more merge of master, and then can merge on green?

You are planning to do a follow-up then to reduce the duplication?

**Comment by simonjayhawkins** [comment]
@jorisvandenbossche green. ok to merge?

**Comment by jorisvandenbossche** [comment]
Yep, thanks!
